Add _voting.py file with voting regressor implementation

doshi-kevin opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs

#26927 
<!--
Example: Fixes  .
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
Includes a link to the plot_voting_regressor.py example.

#### Any other comments?
Please review the code once and correct is any mistakes. I am a beginner and would love to be a contributor of such a large open-source project like scikit-learn.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

doshi-kevin committed:
Add _voting.py file with voting regressor implementation

doshi-kevin committed:
Updates linting error

doshi-kevin committed:
Merge branch 'main' into my_feature

doshi-kevin commented:
Hey @adrinjalali , could you please check this PR and merge this ?

doshi-kevin commented:
@lesteve Could you please check and merge this PR

doshi-kevin committed:
Merge branch 'main' into my_feature

doshi-kevin commented:
Hey @lesteve , could you please tell me what is the error in the PR ?

lesteve commented:
The CI issue is being investigated and hopefully fixed soon in #29771, so it's not related to your PR.

Seems like you put the link to the example at the end of the docstring but I am wondering whether the intent of https://github.com/scikit-learn/scikit-learn/issues/26927 is to put the example link before the parameters for example after the "For more details read the User Guide". Otherwise the link to the example is in the middle of a very verbose page, and IMO this does not really help discoverability ... any thoughs on this @adrinjalali?

You can view the rendering of your PR by clicking on "check the rendered doc link" and then to the class you modified the docstring (`VotingRegressor`) for your last commit: https://output.circle-artifacts.com/output/job/aa86f689-cb7b-4740-851b-f0cb0924a399/artifacts/0/doc/modules/generated/sklearn.ensemble.VotingRegressor.html

lesteve commented:
@doshi-kevin also I am personally not a big fan of being pinged unless this is really blocking and I happen to be the best person to fix something quickly. Different people have different opinions on this.

I would say waiting roughly a week without answer before pinging is probably a reasonable guideline, see for example this FAQ [section](https://scikit-learn.org/stable/faq.html#why-is-my-pull-request-not-getting-any-attention)

doshi-kevin commented:
I  am really sorry, I will make the changes. Sorry for any inconveniences caused.

marenwestermann commented on sklearn/ensemble/_voting.py:
```diff
@@ -637,6 +637,12 @@ class VotingRegressor(RegressorMixin, _BaseVoting):
     >>> len(er.estimators_)
     2
+
+    For a detailed example of utilizing
```
It would be better to be more succinct such as: 

For a more detailed example, refer to :ref:`sphx_glr_auto_examples_ensemble_plot_voting_regressor.py`.

doshi-kevin committed:
Merge branch 'main' into my_feature

doshi-kevin committed:
modified lines

doshi-kevin committed:
Merge remote-tracking branch 'origin/my_feature' into my_feature

doshi-kevin committed:
Merge pull request #1 from doshi-kevin/feature2

Feature2

doshi-kevin committed:
Merge branch 'main' into my_feature

doshi-kevin commented:
Hey @marenwestermann , I checked other similar PRs and I believe I have added the link to the correct location in the file. I added a little text of 'please refer to,'. Please  check this PR, in case of any changes, I will be happy to do it.

doshi-kevin commented:
@marenwestermann Any updates on this ?

adrinjalali commented:
I agree with @marenwestermann's comment here: https://github.com/scikit-learn/scikit-learn/pull/29724/files#r1743929175
The text could be shortened and doesn't need to be long. That single sentence is enough.

@lesteve it depends on whether the example is focusing on the estimator as a whole, or a specific parameter. I believe in this case having it at the end of the docstring is the right place.

doshi-kevin commented:
yes, I will make the text shorter.

doshi-kevin committed:
Merge branch 'scikit-learn:main' into my_feature

doshi-kevin committed:
made the text succint

doshi-kevin committed:
Merge pull request #2 from doshi-kevin/new

made the text succint

adrinjalali commented:
The only thing you need to add is:

> For a more detailed example, refer to :ref:sphx_glr_auto_examples_ensemble_plot_voting_regressor.py.

doshi-kevin commented:
I have shortened the text, please review the changes once :) @adrinjalali @marenwestermann

doshi-kevin committed:
made the text succint

doshi-kevin committed:
Merge branch 'my_feature' into new1

doshi-kevin committed:
Merge pull request #3 from doshi-kevin/new1

made the text succint

doshi-kevin commented:
Hey @adrinjalali , made the changes, sorry for taking too much of your time

Charlie-XIAO commented on sklearn/ensemble/_voting.py:
```diff
@@ -637,6 +637,10 @@ class VotingRegressor(RegressorMixin, _BaseVoting):
     >>> len(er.estimators_)
     2
+
+    For a detailed example of utilizing
+    :class:`~sklearn.ensemble.VotingRegressor`
+    For a more detailed example, refer to :ref:sphx_glr_auto_examples_ensemble_plot_voting_regressor.py.
```
```suggestion
    For a more detailed example, refer to
    :ref:`sphx_glr_auto_examples_ensemble_plot_voting_regressor.py`.
```

You left some redundant lines there, and the formatting and sphinx cross reference syntax is not correct.

doshi-kevin committed:
Update sklearn/ensemble/_voting.py

Co-authored-by: Yao Xiao <108576690+Charlie-XIAO@users.noreply.github.com>

doshi-kevin commented:
Thank you so much @Charlie-XIAO , I have made the changes.

Charlie-XIAO reviewed (commented):
Thanks @doshi-kevin, I'm personally okay with this now, the rendered docs is here: https://output.circle-artifacts.com/output/job/6fbaa7c6-84ad-49bb-8383-9c98a2486271/artifacts/0/doc/modules/generated/sklearn.ensemble.VotingRegressor.html

But indeed as mentioned in https://github.com/scikit-learn/scikit-learn/pull/29724#issuecomment-2328210435 this could have been moved to above parameters, i.e., right after or right before the "read more in user guide" line. I did not follow this meta-issue but it seems there are related PRs doing in both ways. For this specific case it's probably better to move it upwards (I wonder if one can find where the link is from the screenshot below at first glance 🤔)

![image](https://github.com/user-attachments/assets/3c169636-7000-421b-b98a-01fa5ad3cf5c)

doshi-kevin committed:
Merge branch 'main' into my_feature

doshi-kevin commented:
@Charlie-XIAO Noted. I will move the link upwards (above "read more in user guide" line)

doshi-kevin committed:
changes link position

doshi-kevin committed:
Merge branch 'my_feature' into new2

doshi-kevin committed:
Merge pull request #4 from doshi-kevin/new2

changes link position

doshi-kevin commented:
@Charlie-XIAO @adrinjalali I have made the change and put the link in the beginning of the page above ...see more User Guide. Please check once. Thank You.

Charlie-XIAO commented on sklearn/ensemble/_voting.py:
```diff
@@ -555,6 +555,9 @@ class VotingRegressor(RegressorMixin, _BaseVoting):
     regressors, each on the whole dataset. Then it averages the individual
     predictions to form a final prediction.
 
+    For a more detailed example, refer to
```
```suggestion
    For a detailed example, refer to
```

Since we moved it up here the wording needs to be changed a bit.

Charlie-XIAO reviewed (approved):
LGTM thanks, just one nitpick below:

doshi-kevin committed:
Update sklearn/ensemble/_voting.py

Co-authored-by: Yao Xiao <108576690+Charlie-XIAO@users.noreply.github.com>

Charlie-XIAO commented:
Everything looks good now, thanks @doshi-kevin, I'm enabling auto-merge.
